src/groupNoble.ts

```diff
@@ -0,0 +1,337 @@
+        return this.p.equals(a.p)
+    }
+
+    neg(): EltNb {
+        return new EltNb(this.g, this.p.negate())
+    }
+
+    add(a: EltNb): EltNb {
+        compat(this, a)
+        return new EltNb(this.g, this.p.add(a.p))
+    }
+
+    mul(s: ScalarNb): EltNb {
+        compat(this, s)
+        return new EltNb(this.g, this.p.multiply(s.k))
+    }
+
+    mul2(k1: ScalarNb, a: EltNb, k2: ScalarNb): EltNb {
+        compat(this, k1)
+        compat(this, k2)
+        compat(this, a)
+        const el = this.p.multiplyAndAddUnsafe(a.p, k1.k, k2.k)
+        if (!el) throw new Error('result is zero')
+        return new EltNb(this.g, el)
+    }
+
+    serialize(compressed = true): Uint8Array {
+        if (this.isIdentity()) {
+            return Uint8Array.from([0])
+        }
+        return this.p.toRawBytes(compressed)
+    }
+
+    // size returns the number of bytes of a non-zero element in compressed or uncompressed form.
+    static size(g: GroupNb, compressed = true): number {
+        return 1 + (compressed ? g.size : g.size * 2)
+    }
+
+    private static deser(g: GroupNb, bytes: Uint8Array): EltNb {
+        const point = g.params.Point.fromHex(bytes)
+        point.assertValidity()
+        return new EltNb(g, point)
+    }
+
+    // Deserializes an element, handles both compressed and uncompressed forms.
+    static deserialize(g: GroupNb, bytes: Uint8Array): EltNb {
+        const len = bytes.length
+        switch (true) {
+            case len === 1 && bytes[0] === 0x00:
+                return g.identity()
+            case len === 1 + g.size && (bytes[0] === 0x02 || bytes[0] === 0x03):
+                return EltNb.deser(g, bytes)
+            case len === 1 + 2 * g.size && bytes[0] === 0x04:
+                return EltNb.deser(g, bytes)
+            default:
+                throw errDeserialization(EltNb)
+        }
+    }
+
+    static hash(g: GroupNb, msg: Uint8Array, dst: Uint8Array): EltNb {
+        const h2c = g.params.h2c
+        const p = h2c(msg, { DST: decodeDST(dst) }) as ProjPointType<bigint>
+        return new EltNb(g, p)
+    }
+}
+
+class GroupNb implements Group {
+    static readonly Elt: EltCons = EltNb
+    static readonly Scalar: ScalarCons = ScalarNb
+    static readonly ID = GroupIDs
+    static for(gid: GroupID) {
+        return new this(gid)
+    }
+
+    public readonly params: GroupParams
+    public readonly size: number
+    public readonly id: GroupID
+
+    constructor(gid: GroupID) {
+        this.params = getParams(gid)
+        this.size = this.params.size
+        this.id = gid
+    }
+
+    newScalar(): ScalarNb {
+        return ScalarNb.new(this)
+    }
+
+    newElt(): EltNb {
+        return this.identity()
+    }
+
+    identity(): EltNb {
+        return EltNb.new(this)
+    }
+
+    generator(): EltNb {
+        return EltNb.gen(this)
+    }
+
+    mulGen(s: ScalarNb): EltNb {
+        return EltNb.gen(this).mul(s)
+    }
+
+    async randomScalar(): Promise<ScalarNb> {
```

Comment:
Some linters actually enforce consistent use of async, which is something I can get behind, but if that's the convention here, I can go with it. Maybe there's a linting rule this project could set up ?

Reply:
async randomScalar(): Promise<ScalarNb> {
        const msg = crypto.getRandomValues(new Uint8Array(this.size))
        return ScalarNb.hash(this, msg, new Uint8Array())
    }

Actually, looking at this more, I can see that the async modifier is what makes the simple `return value` upgrade to a Promise, rather than having to use Promise.resolve(value) 

Our Group interfaces are async (using Promises), so the async is actually required?

Reply:
until there is a linter rule enabled to get this uniformly, I consider that using `async` only when inside the body of the function there is an `await` keyword.

A function is known to be async iff it returns a `Promise`.

Reply:
> A function is known to be async iff it returns a Promise.

Let's dig into this a bit more. 

The randomScalar method of our Group interface is asynchronous so that it can support the sjcl/webcrypto (used for expandMD) implementation. 

> scalarNb.hash(this, msg, new Uint8Array())

This is a sync operation, and not a promise, it's true, but by putting `async` on the function it "upgrades" the return value to a Promise so it will conform to our interface.

The alternative here is to use Promise.resolve(), which I will do: 

```typescript
    randomScalar(): Promise<ScalarNb> {
        const msg = crypto.getRandomValues(new Uint8Array(this.size))
        return Promise.resolve(ScalarNb.hash(this, msg, new Uint8Array()))
    }
```